File: java/src/main/java/dev/sanmer/sac/io/Library.java
package dev.sanmer.sac.io;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;

public class Library {
    private static final String LIBRARY_NAME = "libsac-jni.dylib";

    public static void load() {
        String libraryFile = getLibraryFile();
        System.load(libraryFile);
    }

    private static String getLibraryFile() {
        String sacJniLibraryPath = System.getenv("SAC_JNI_LIBRARY_PATH");
        if (sacJniLibraryPath != null) {
            return sacJniLibraryPath;
        }

        URL libraryURL = Library.class.getResource("/" + LIBRARY_NAME);
        if (libraryURL == null || !libraryURL.getProtocol().equals("jar")) {
            throw new LinkageError("Load " + LIBRARY_NAME);
        }

        File libsDir;
        try {
            Path tempDir = Files.createTempDirectory("sac-jni");
            tempDir.toFile().deleteOnExit();
            libsDir = tempDir.toFile();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        File libraryFile = getFile(libsDir, libraryURL);

        return libraryFile.getAbsolutePath();
    }

    @SuppressWarnings("resource")
    @NotNull
    private static File getFile(File libsDir, URL libraryURL) {
        File libraryFile = new File(libsDir, LIBRARY_NAME);
        libraryFile.deleteOnExit();

        try {
            InputStream inputStream = libraryURL.openStream();
            FileOutputStream outputStream = new FileOutputStream(libraryFile);

            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                outputStream.write(buffer, 0, bytesRead);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return libraryFile;
    }
}

File: src/commonMain/kotlin/dev/sanmer/sac/io/Library.kt
package dev.sanmer.sac.io

import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files

internal interface Library {
    val name: String
    fun load()

    companion object {
        private const val PATH = "LIBSAC_JNI_PATH"
        private const val TMP = "sac-jni"

        internal fun <T: Library> T.getLibrary(): String {
            System.getenv(PATH)?.let { path ->
                return path
            }

            val library = javaClass.getResource("/${name}")
            if (library?.protocol != "jar") {
                throw LinkageError("Load $name")
            }

            val libsDir = Files.createTempDirectory(TMP).toFile()
            val libraryFile = File(libsDir, name)

            val inputStream = library.openStream()
            val outputStream = FileOutputStream(libraryFile)

            inputStream.use { input ->
                outputStream.use { output ->
                    input.copyTo(output)
                }
            }

            return libraryFile.absolutePath
        }
    }
}